sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -617,13 +617,18 @@ class Analyzer(
     private def lookupTableFromCatalog(
         u: UnresolvedRelation,
         defaultDatabase: Option[String] = None): LogicalPlan = {
+      val db = u.tableIdentifier.database.orElse(defaultDatabase)
+      val tableIdentWithDb = u.tableIdentifier.copy(database = db)
       try {
-        val tableIdentWithDb = u.tableIdentifier.copy(
-          database = u.tableIdentifier.database.orElse(defaultDatabase))
         catalog.lookupRelation(tableIdentWithDb, u.alias)
       } catch {
         case _: NoSuchTableException =>
-          u.failAnalysis(s"Table or view not found: ${u.tableName}")
+          u.failAnalysis(s"Table or view not found: ${tableIdentWithDb.unquotedString}")
+        // If the database is defined and that database is not found, throw an AnalysisException.
+        // Note that if the database is not defined, it is possible we are looking up a temp view.
+        case _: NoSuchDatabaseException if db.isDefined =>
```

Comment:
if users don't specify database, and the current database is removed by other sessions, then they will hit `NoSuchDatabaseException`. In this case, I think we should also throw `AnalysisException` for consistency.